### ISSUE TITLE
Static dictionary causing exception

### ISSUE THREAD

**Issue by phillip-haydon** [aws/aws-sdk-net#986]
https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/RegionEndpoint/RegionEndpoint.cs#L41

On occasion, I'm getting exceptions caused by this being a static dictionary that isn't thread safe.

```
Message: An item with the same key has already been added.
     System.ThrowHelper.ThrowArgumentException(ExceptionResource resource):16
     System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add):0
     Amazon.RegionEndpoint.NewEndpoint(String systemName, String displayName):25
     Amazon.RegionEndpoint.GetBySystemName(String systemName):72
```

It doesn't happen often, maybe 1-2 times a month, mostly after a deployment if there's a bit of activity going on when the deployment happens, so when it spins up it seems 2 different threads try to create S3 client, and they fall into this exception.

```
            var config = string.IsNullOrEmpty(region)
                ? new AmazonS3Config { ServiceURL = awsUrl }
                : new AmazonS3Config { RegionEndpoint = RegionEndpoint.GetBySystemName(region) };

            amazonClient = AWSClientFactory.CreateAmazonS3Client(user, password, config);
```

This is the code that causes the issue.

I think the code could be updated to use a ConcurrentDictionary?

**Comment by normj** [comment]
Interesting, are you using a newer region then your version version of AWSSDK.Core knows about? Seems like all of the known regions that have a constant would not cause this problem but I can see if 2 threads are using a region that your version of Core didn't know about it could cause this problem as it dynamically add the unknown region to the collection.

**Comment by phillip-haydon** [comment]
I just realised we are using the older AWSSDK v2.3.55.2.

Maybe the issue is with that, I'll see if we can update to V3.

**Comment by phillip-haydon** [comment]
I'm unable to reproduce the error using V3, so I'll close this. Thanks.

**Comment by phillip-haydon** [comment]
@normj 

It appears the bug still occurs: 

> Message: An item with the same key has already been added.

```csharp
System.ThrowHelper.ThrowArgumentException(ExceptionResource resource):16
System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add):129
Amazon.RegionEndpoint.NewEndpoint(String systemName, String displayName):25
Amazon.RegionEndpoint.GetBySystemName(String systemName):17
```

I'm adding a try/catch around our usage of the SDK here, but I think the SDK needs to be more defensive, if possible change it to a concurrent dictionary.

**Comment by stilettk** [comment]
Can confirm, the issue happens with the new eu-north-1 region because it is still not added to .NET SDK.

**Comment by stilettk** [comment]
It seems that the internal dictionary is initialized with regions from static properties. Since eu-north-1 and us-gov-east-1 static properties are still not added, I have found a workaround to avoid this error. You need to explicitly add new regions to the dictionary by calling `RegionEndpoint.GetBySystemName`:
```C#
// RegionEndpoint class is not thread-safe.
// We need to load new regions explicitly to avoid errors later 
RegionEndpoint.GetBySystemName("eu-north-1");
RegionEndpoint.GetBySystemName("us-gov-east-1");
```
Please note that you wil need to do this for all future regions until their corresponding static properties will be added (or until this bug will be fixed). For example, for us-gov-east-1 it's 1.5 months now and counting.

**Comment by ckuehl** [comment]
Related to this, when can we expect to see `eu-north-1` added to `aws-sdk-net/sdk/src/Core/RegionEndpoint/RegionEndpoint.cs`?

**Comment by klaytaybai** [comment]
@ckuehl, thank you for inquiring about the new eu-north-1 region.  I am not aware of anything preventing us from adding this and other regions that are still pending, but there may be.  I will discuss it with the team later today.  If there isn't anything we are waiting on, you could probably expect it to go out in the next release or one shortly thereafter.  Because it is somewhat unrelated to this issue, please create a new issue if you would like it tracked and given updates.  Thanks.

**Comment by ckuehl** [comment]
@klaytaybai Done - created #1167.

**Comment by ckuehl** [comment]
We are still seeing the original issue reported here using a recent version of the SDK:

```
System.ArgumentException: An item with the same key has already been added. Key: ap-east-1
  Module "System.ThrowHelper", in ThrowAddingDuplicateWithKeyArgumentException
  Module "System.Collections.Generic.Dictionary`2", in TryInsert
  Module "Amazon.RegionEndpoint", in NewEndpoint
  Module "Amazon.RegionEndpoint", in get_EnumerableAllRegions
```

**Comment by ckuehl** [comment]
We just saw a similar/related issue in production on one of our machines (but not all):

```
System.ArgumentException: An item with the same key has already been added. Key: System.Collections.Generic.HashSet`1[System.String]
at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
at Amazon.Internal.RegionEndpointV3.ServiceMap.Add(String serviceName, Endpoint endpoint, HashSet`1 variants)
at Amazon.Internal.RegionEndpointV3.AddVariantEndpointsToServiceMap(JsonData mergedEndpoint, String regionName, String serviceName, Dictionary`2 mergedVariants)
at Amazon.Internal.RegionEndpointV3.AddServiceToMap(JsonData service, String serviceName)
at Amazon.Internal.RegionEndpointV3.ParseAllServices()
at Amazon.Internal.RegionEndpointV3.GetEndpointForService(String serviceName, HashSet`1 variants)
at Amazon.Internal.RegionEndpointV3.GetEndpointForService(String serviceName, Boolean dualStack)
at Amazon.Runtime.ClientConfig.GetUrl(RegionEndpoint regionEndpoint, String regionEndpointServiceName, Boolean useHttp, Boolean useDualStack)
at Amazon.Runtime.ClientConfig.DetermineServiceURL()
at Amazon.Runtime.Internal.DefaultRetryPolicy..ctor(IClientConfig config)
at Amazon.Runtime.AmazonServiceClient.BuildRuntimePipeline()
```

Can someone please investigate?

**Comment by ckuehl** [comment]
@normj can you please assist with re-opening this?
